Add rainbow indentation guides

omentic opened the pull request:
Supercedes #4056 and #4493. Closes #4010.

This PR is a rebase of #4493 now that #13530 has been merged. It adds a `rainbow-indent` option to the `[editor.indent-guides]` configuration, taking three values: `"none"` (default), `"dim"`, or `"normal"`.

Rendered (normal, dim):
<img width="auto" height="200" alt="2025-07-29-025824" src="https://github.com/user-attachments/assets/27eb728c-f905-47c8-971c-52a8a0a590ef" /> <img width="auto" height="200" alt="2025-07-29-025828" src="https://github.com/user-attachments/assets/a3f8b3ed-4446-4eef-b73f-57eb71250b2a" />

SoraTenshi committed:
Add rainbow indentation guides

omentic commented:
Let me know if there's anything to change. Should "normal" be called "bright"?

yerlaser commented:
I just tried it with "dim" setting.
The lines look darker and colors are harder to distinguish.
Is this because of my terminal or is it by design?

Otherwise, it's a cool addition. Just "normal" looks a bit too bright indeed.

yerlaser commented:
Ideally the lines would be dotted lines but I understand it's impossible in terminal.

mi2ebi commented:
helix lets you change what character is used for indent guides

DerSaidin commented on book/src/editor.md:
```diff
@@ -321,11 +321,12 @@ tabpad = "·" # Tabs will look like "→···" (depending on tab width)
+| `render`         | Whether to render indent guides                           | `false` |
+| `character`      | Literal character to use for rendering the indent guide   | `│`     |
+| `skip-levels`    | Number of indent levels to skip                           | `0`     |
+| `rainbow-indent` | Enum to set rainbow indentations. `normal`, `dim`, `none` | `none`  |
```
+1 for just a bool and rely on themes

-0.1 for `rainbow` for both. 2 options is simpler than 3. explicitly configuring both is not hard.

DerSaidin commented on book/src/themes.md:
```diff
@@ -132,7 +132,7 @@ berry = "#2A2A4D"
 ### Rainbow
 
-The `rainbow` key is used for rainbow highlight for matching brackets.
+The `rainbow` key is used for rainbow highlight for matching brackets and indentation.
```
Should the rainbow for indentation us a separate key?

omentic replied:
No. See here: https://github.com/helix-editor/helix/pull/4056#discussion_r995492132

omentic committed:
Refactor rainbow indent guides configuration into being a boolean on/off

omentic commented:
I've refactored this PR to expose a boolean `editor.indent-guides.rainbow` configuration option, rather than the previous string `editor.indent-guides.rainbow-indent` option. The options of dim / bright indent guides are moved to theme-side, which exposes new `rainbow-brackets` and `rainbow-indents` keys for theming, that override the default `rainbow` key.

I would appreciate if someone more familiar with the codebase would have a look. While I've tested this a little bit and haven't hit any issues, there are some parts of the codebase I've now touched that I do not understand. In particular:
- [ ] Where the names of styles (within `build_theme_values` in `helix-view/src`) are read / what they are used for is unclear to me. Since we're now pushing separate bracket and indent guide themes, I've renamed the names from `"rainbow.{i}"` to `"rainbow-bracket.{i}"` and `"rainbow-indent.{i}"` accordingly.
- [ ] Where the styles pushed to `highlights` in that function are accessed is unclear to me. I would think you would have needed to index by an offset of `rainbow_length`, and I went looking to replace it with `rainbow_bracket_length + rainbow_indent_length` but... could not find any such place.
- [ ] What `doc_rainbow_highlights()` in `helix-term/src/ui/editor.rs` is doing is unclear to me. I briefly touched it to pass in `theme.rainbow_bracket_length()`. I do not know if this is correct.
- [ ] What `rainbow_highlights()` in `helix-core/src/syntax.rs` is doing is unclear to me. I didn't need to touch it, so I didn't.

poliorcetics commented on book/src/themes.md:
```diff
@@ -141,6 +141,20 @@ rainbow = ["#ff0000", "#ffa500", "#fff000", { fg = "#00ff00", modifiers = ["bold
 
 Colors from the palette and modifiers may be used.
 
+The `rainbow-brackets` and `rainbow-indents` keys are used for rainbow highlights for matching brackets and indentation guides, respectively.
```
Helix automatically fall back it theme keys are of the format `"outer.middle.inner" = ...` (dot-separated), you should use that instead :)

phizlip commented:
+1 for getting this merged

code-instable commented:
this would be very useful indeed

yerlaser commented:
Doesn't build any more when rebased on master.
I resolved conflicts in my unstable fork as per my understanding of the code.
However, would be very glad if you rebase it once for us.

omentic commented:
Sure, I'll look at rebasing soon -- been busy, but will probably have some time next week.
